Add support for Self type

ilevkivskyi opened the pull request:
Ref #12840 
Fixes #11871
Fixes #14089

This is an alternative implementation to two existing PRs: https://github.com/python/mypy/pull/11666, https://github.com/python/mypy/pull/13133. This PR treats `typing.Self` as pure syntactic sugar, and transforms it into a type variable early during semantic analyzis.

This way we can re-use all the existing machinery and handled edge cases for self-types. The only new thing is self-type for _attributes_ (as proposed in the PEP). This required handling in several places, since attribute access is duplicated in several places (see #7724), plus special forms (like NamedTuples and TypedDicts) and dataclasses plugin require additional care, since they use attribute annotations in special ways.

I don't copy all the existing tests for "old style" self-types, but only some common use cases, possible error conditions, and relevant new edge cases, such as e.g. special forms mentioned above, and implicit type variable binding for callable types.

cc @Gobot1234 (author of original PRs)

Gobot1234 commented:
Looks great. One test that's been missed is ClassVar[Self] which will currently fail as I understand it (see https://github.com/python/mypy/pull/13133/files#diff-f782aad4853edceb67a88efe43d2869579d075a77b8e8dd2add91e818a9fa164R4048-R4051)

ilevkivskyi commented:
Yeah, I wasn't sure about `ClassVar`, and it is not covered in the PEP, so I just kept whatever semantics we have. It is probably easy to tweak, if people want to use it. I guess possible use case is various "singleton" or "canonical" instances, for example:
```python
class Context:
    BACKGROUND: ClassVar[Self]
    ...

Context.BACKGROUND = Context()
```

ilevkivskyi commented:
@Gobot1234 Hm, OK. But also looking at the original example in that issue I am not use we should allow `Self` in other unusual places (i.e. those not covered by [Valid Locations for Self](https://peps.python.org/pep-0673/#valid-locations-for-self) section). They work inconsistently in current version of this PR, I wasn't really careful (although they are probably not hard to support).

Also I will need to disable `Self` in metaclasses, as specified by the PEP.

@hauntsaninja Is there a way to get non-truncated output of `mypy_primer`? (if I don't want to run it locally :-))

ilevkivskyi commented on mypy/semanal.py:
```diff
@@ -926,10 +938,21 @@ def prepare_method_signature(self, func: FuncDef, info: TypeInfo) -> None:
                         leading_type = self.class_type(leading_type)
                     func.type = replace_implicit_first_type(functype, leading_type)
+                elif has_self_type and isinstance(func.unanalyzed_type, CallableType):
+                    if not isinstance(get_proper_type(func.unanalyzed_type.arg_types[0]), AnyType):
+                        self.fail(
+                            "Method cannot have explicit self annotation and Self type", func
+                        )
```
Yeah, actually my idea was to prohibit _unrelated_ annotation on `self` (see tests). I will special case this check to allow actual `Self` type (or maybe give it a special error code, like `[redundant-self]`).

ilevkivskyi replied:
I mean this part from the PEP
```python
class Foo:
    # Rejected (Self is treated as unknown).
    def has_existing_self_annotation(self: T) -> Self: ...
```
Btw while looking at the PEP it looks like I also forgot to prohibit `Self[...]` (with type arguments). Currently they are just silently ignored.

JukkaL commented on docs/source/generics.rst:
```diff
@@ -335,6 +328,43 @@ possibly by making use of the ``Any`` type.
+.. note::
+
+   This feature is available on Python 3.11 or newer. On older Python versions
+   you can import backported ``Self`` from latest ``typing_extensions``.
```
I'd suggest replacing 'latest' with 'recent enough' since otherwise it go stale quickly.

JelleZijlstra replied:
Also, it's misleading to start with "This feature is available on Python 3.11 or newer." Users will think they can't use `Self` before 3.11. I'd suggest saying:
```suggestion
   ``Self`` can be imported from ``typing`` on Python 3.11 or higher, and also from
   ``typing_extensions`` on version 4.0.0 or higher.
```
(Ref: https://github.com/python/typing_extensions/blob/main/CHANGELOG.md#release-400-november-14-2021)

ilevkivskyi replied:
Yes, I will make it more clear (or maybe just commit what Jelle proposed)

JukkaL commented on mypy/checkmember.py:
```diff
@@ -945,7 +951,12 @@ def analyze_class_attribute_access(
+            if node.node.info.self_type is not None and node.node.is_classvar:
+                exclude = {node.node.info.self_type.id}
+            else:
+                exclude = set()
```
Optional optimization: this code path may be somewhat hot? Replacing a set with zero or one items with an optional type would be faster. You can use `!=` instead of ` not in`, I think.

JukkaL commented on mypy/semanal_shared.py:
```diff
@@ -324,3 +331,11 @@ def visit_placeholder_type(self, t: PlaceholderType) -> bool:
     return typ.accept(HasPlaceholders())
+
+
+def special_self_type(info: TypeInfo | None, defn: Context) -> Type:
```
Add docstring?

JukkaL commented on test-data/unit/check-dataclasses.test:
```diff
@@ -1932,3 +1932,25 @@ B = List[C]
+@dataclass
+class LinkedList(Generic[T]):
+    value: T
+    next: Optional[Self] = None
```
Test accessing `self.next` in a dataclass method?

JukkaL commented on test-data/unit/check-protocols.test:
```diff
@@ -3843,3 +3843,23 @@ def f() -> str: ...
+
+z: P
+reveal_type(S().next)  # N: Revealed type is "__main__.S"
+reveal_type(z.next)  # N: Revealed type is "__main__.P"
```
Maybe also test a subprotocol of `P` and accessing inherited property?

ilevkivskyi replied:
I added subprotocol test (didn't push yet). For inherited property there is `testTypingSelfProperty` in a different test file.

JukkaL commented on test-data/unit/check-selftype.test:
```diff
@@ -1267,3 +1267,206 @@ class Test(Generic[T]):
+    x: Self
+
+class D(C[int]):
+    x: D
```
What if we now have a class like this:
```py
class E(D): pass
```
This should perhaps be an error (incompatible attribute), since the type of `x` is `D` in `E`, which could be considered incompatible with the definition in `C`. It looks like we don't have a similar check for methods, so it's okay to not generate an error.

ilevkivskyi replied:
Yeah, this looks wrong, but also may be not super easy to implement. Let's postpone this for now.

hauntsaninja replied:
Worth adding an xfail test?

ilevkivskyi replied:
TBH it is unlikely it will be fixed accidentally, instead I will open a follow up issue, maybe at some point we will have energy to tighten this.

JukkaL commented on test-data/unit/check-typeddict.test:
```diff
@@ -2590,3 +2590,30 @@ TD[str](key=0, value=0)  # E: Incompatible types (expression has type "int", Typ
+T = TypeVar("T")
+class TD(TypedDict, Generic[T]):
+    val: T
+    next: Optional[Self]
```
What happens if we inherit a typed dict from `TD`?

ilevkivskyi replied:
TypedDicts can't be inherited in a "normal" sense of this word, since we can't change item types, we can only add new items, so I usually call it extending.

Currently I eagerly expand `Self` in TypedDicts (and in NamedTuples), so in this case `next` will be still `Optional[TD]` in extended TypedDict. PEP doesn't say anything about TypedDicts and NamedTuples, and they are not really classes, so I just implemented whatever looked more natural (and also simpler to implement).

But actually thinking a bit more about this, maybe it would be better to implement them more like for classes, so it will point to extended dict. I am not sure yet.

JukkaL commented on test-data/unit/check-selftype.test:
```diff
@@ -1267,3 +1267,206 @@ class Test(Generic[T]):
+
+class C:
+    attr: List[Self]
+    def meth(self) -> List[Self]: ...
```
Can you also test returning `self` vs. returning `C()` inside a method that returns `Self` (the latter should be an error).

JukkaL commented on docs/source/generics.rst:
```diff
@@ -264,15 +264,8 @@ Generic methods and generic self
-
-   This feature is experimental. Checking code with type annotations for self
-   arguments is still not fully implemented. Mypy may disallow valid code or
-   allow unsafe code.
```
I think that we still allow some unsafe code. It would be good to document some cases if we remove this note. I can give some examples if it's helpful.

JukkaL reviewed (commented):
Nice! Left some comments (not a full review).

gandhis1 commented:
As a passerby - is there any way to view the truncated output of mypy primer?

ilevkivskyi commented:
@JukkaL Thanks for comments! I addressed almost all of them.

I am not sure what to do with the uses in TypedDicts (and also in NamedTuples). Btw PEP actually prohibits `Self` in type aliases because it is ambiguous. Maybe we should also prohibit `Self` in TypedDicts and NamedTuples as well? They also can be quite ambiguous, especially when using call syntax, and even more when using call syntax inside another class.

ilevkivskyi commented:
Hm, it looks like `mypy_primer` shows there is a bug somewhere: unrelated TypeVars end up with same id, causing a confusion. I can't find a simple repro yet.

JelleZijlstra commented on mypy/semanal.py:
```diff
@@ -954,15 +1012,37 @@ def f(): ...  # Error: 'f' redefined
         """Make any type variables in the signature of defn explicit.
 
         Update the signature of defn to contain type variable definitions
         if defn is generic.
```
Update the docstring to say what the return value means.

JelleZijlstra reviewed (commented):
Thanks for implementing this!

yakMM commented:
Hello, thanks for developing this feature, I was testing your branch and I am unable to make Self work for classmethods, is it expected behavior?

```py
import typing as t

class Foo:
    @classmethod
    def bar(cls) -> t.Self:
        return cls()
```

```
PS F:\jim> pdm run mypy .\main.py --show-traceback      
main.py:7: error: INTERNAL ERROR -- Please try using mypy master on GitHub:
https://mypy.readthedocs.io/en/stable/common_issues.html#using-a-development-mypy-build
Please report a bug at https://github.com/python/mypy/issues
version: 1.0.0+dev.61c05895fac6da6d19e144b3540b173ae46dfbbc
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "F:\jim\__pypackages__\3.11\Scripts\mypy.EXE\__main__.py", line 7, in <module>
    sys.exit(console_entry())
  File "F:\jim\__pypackages__\3.11\lib\mypy\__main__.py", line 15, in console_entry
    main()
  File "F:\jim\__pypackages__\3.11\lib\mypy\main.py", line 95, in main
    res, messages, blockers = run_build(sources, options, fscache, t0, stdout, stderr)
  File "F:\jim\__pypackages__\3.11\lib\mypy\main.py", line 174, in run_build
    node.accept(self)
  File "F:\jim\__pypackages__\3.11\lib\mypy\nodes.py", line 834, in accept
    return visitor.visit_func_def(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "F:\jim\__pypackages__\3.11\lib\mypy\semanal.py", line 814, in visit_func_def
    self.analyze_func_def(defn)
  File "F:\jim\__pypackages__\3.11\lib\mypy\semanal.py", line 871, in analyze_func_def
    self.analyze_function_body(defn)
  File "F:\jim\__pypackages__\3.11\lib\mypy\semanal.py", line 1327, in analyze_function_body
    a.bind_function_type_variables(cast(CallableType, defn.type), defn)
  File "F:\jim\__pypackages__\3.11\lib\mypy\typeanal.py", line 1379, in bind_function_type_variables
    assert var_node, "Binding for function type variable not found within function"
AssertionError: Binding for function type variable not found within function
main.py:7: : note: use --pdb to drop into pdb

```

yakMM commented:
Hello, thanks a lot for developing this feature. I was testing it out and realized it's only working when importing `Self` explicitly.

For example this is fine:

```py
from typing import Self


class Foo:
    @classmethod
    def bar(cls) -> Self:
        return cls()

```

But this is not:
```py
import typing as t


class Foo:
    @classmethod
    def bar(cls) -> t.Self:
        return cls()
```

```
PS F:\jim> pdm run mypy .\main.py --show-traceback
main.py:6: error: INTERNAL ERROR -- Please try using mypy master on GitHub:
https://mypy.readthedocs.io/en/stable/common_issues.html#using-a-development-mypy-build
Please report a bug at https://github.com/python/mypy/issues
version: 1.0.0+dev.61c05895fac6da6d19e144b3540b173ae46dfbbc
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "F:\jim\__pypackages__\3.11\Scripts\mypy.EXE\__main__.py", line 7, in <module>
    sys.exit(console_entry())
  File "F:\jim\__pypackages__\3.11\lib\mypy\__main__.py", line 15, in console_entry
    main()
  File "F:\jim\__pypackages__\3.11\lib\mypy\main.py", line 95, in main
    res, messages, blockers = run_build(sources, options, fscache, t0, stdout, stderr)
  File "F:\jim\__pypackages__\3.11\lib\mypy\main.py", line 174, in run_build
    res = build.build(sources, options, None, flush_errors, fscache, stdout, stderr)
  File "F:\jim\__pypackages__\3.11\lib\mypy\build.py", line 193, in build
    result = _build(
  File "F:\jim\__pypackages__\3.11\lib\mypy\build.py", line 276, in _build
    graph = dispatch(sources, manager, stdout)
  File "F:\jim\__pypackages__\3.11\lib\mypy\build.py", line 2903, in dispatch
    process_graph(graph, manager)
  File "F:\jim\__pypackages__\3.11\lib\mypy\build.py", line 3287, in process_graph
    process_stale_scc(graph, scc, manager)
  File "F:\jim\__pypackages__\3.11\lib\mypy\build.py", line 3382, in process_stale_scc
    mypy.semanal_main.semantic_analysis_for_scc(graph, scc, manager.errors)
  File "F:\jim\__pypackages__\3.11\lib\mypy\semanal_main.py", line 85, in semantic_analysis_for_scc
    process_functions(graph, scc, patches)
  File "F:\jim\__pypackages__\3.11\lib\mypy\semanal_main.py", line 243, in process_functions
    process_top_level_function(
  File "F:\jim\__pypackages__\3.11\lib\mypy\semanal_main.py", line 282, in process_top_level_function
    deferred, incomplete, progress = semantic_analyze_target(
  File "F:\jim\__pypackages__\3.11\lib\mypy\semanal_main.py", line 339, in semantic_analyze_target
    analyzer.refresh_partial(
  File "F:\jim\__pypackages__\3.11\lib\mypy\semanal.py", line 587, in refresh_partial
    self.accept(node)
  File "F:\jim\__pypackages__\3.11\lib\mypy\semanal.py", line 6139, in accept
    node.accept(self)
  File "F:\jim\__pypackages__\3.11\lib\mypy\nodes.py", line 834, in accept
    return visitor.visit_func_def(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "F:\jim\__pypackages__\3.11\lib\mypy\semanal.py", line 814, in visit_func_def
    self.analyze_func_def(defn)
  File "F:\jim\__pypackages__\3.11\lib\mypy\semanal.py", line 871, in analyze_func_def
    self.analyze_function_body(defn)
  File "F:\jim\__pypackages__\3.11\lib\mypy\semanal.py", line 1327, in analyze_function_body
    a.bind_function_type_variables(cast(CallableType, defn.type), defn)
  File "F:\jim\__pypackages__\3.11\lib\mypy\typeanal.py", line 1379, in bind_function_type_variables
    assert var_node, "Binding for function type variable not found within function"
AssertionError: Binding for function type variable not found within function
```

Is it expected behavior?

ilevkivskyi commented:
> Is it expected behavior?

No, this is just a bug I am going to fix.

ilevkivskyi commented:
OK, I have addressed all CR, and I am happy with `mypy_primer`. I also added two follow up issues:
* https://github.com/python/mypy/issues/14074
* https://github.com/python/mypy/issues/14075

If there will be no more comments, I will merge this in next couple days.

ilevkivskyi commented:
It looks like there are no more comments, I will be merging this later today.

patrick91 commented:
@ilevkivskyi this hasn't been release yet, right?

hauntsaninja commented:
No, not yet

henryiii commented:
Now it has. :)
